doc/10_Rendering_Projects/00_Configuration_Reference.md
## Project Configuration Reference
The following is the generated reference for the `mds_pim_print_core` configuration tree. This reference can be generated by issuing the following command:
```bash
bin/console config:dump-reference mds_pim_print_core
```
```yaml
# Default configuration for extension with alias: "mds_pim_print_core"
mds_pim_print_core:

    # Optional locales to set in render mode to have float values converted to '.-strings'
    lc_numeric:

        # Examples:
        - en_US.UTF-8
        - de_DE.UTF-8

    # Optional host settings for proxy or multi domain environments.
    host:

        # Hostname
        hostname:             ~

        # Protocol
        protocol:             ~

        # Port
        port:                 ~
    projects:

        # Prototype
        -

            # Internal project identifier.
            ident:                ~

            # Name of project displayed in InDesign-Plugin.
            name:                 ~ # Required

            # Service to use to render the project. Must inherit AbstractProject.
            service:              ~ # Required

            # Toggles creation of update info layers.
            create_update_info:   true

            # InDesign Template settings.
            template:             # Required

                # Default InDesign template filename.
                default:              ~

                # Optional relative path inside service bundle to the InDesign template.
                relative_path:        /Resources/pimprint/

                # Download template
                download:             true

            # Available plugin elements.
            plugin_elements:

                # Field for update modes.
                update_mode:          true

                # Available update modes for project.
                update_modes:

                    # Defaults:
                    - 501
                    - 502
                    - 512

                # Field for start left/right page.
                start_alignment:      false

                # Fields for page start/end.
                page_bounds:          false

            # Asset handling settings.
            assets:

                # Toggles asset download.
                download:             true

                # Toggles asset download before rendering start.
                pre_download:         true

                # Toggles missing asset onPage warnings messages.
                warnings_on_page:     true

                # Optional Pimcore thumbnail configuration for preview images.
                thumbnail:            ~

            # Optional PHP setting time_limit for project generation.
            php_time_limit:       '0'

            # Optional PHP setting memory_limit for project generation.
            php_memory_limit:     2G
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.2.1] - 2022-04-25
### Fix
-  [[AbstractProject] Loading languages in multi-domain setup #1](https://github.com/mds-agenturgruppe/pimprint-core-bundle/pull/1)
  - Uses `\Pimcore\Localization\LocaleService` instead of `\Mds\PimPrint\CoreBundle\Service\UserHelper`

### Breaking changes
- Removed `\Mds\PimPrint\CoreBundle\Service\UserHelper`

## [1.2.0] - 2020-10-22
### Features
- InDesign template file download with PimPrint-Plugin.
- Simplified API for content aware updates.
- `PublicationTreeBuilder` service for direct usage.
- `Template` commands for single page and facing page documents.
- `SplitTable` command for automatic table splitting across multiple pages.

### Breaking changes
- Change `GoToPage` and `NextPage` constructor signature.
- Rename namespace to `Mds\PimPrint\CoreBundle\InDesign\Command\Variables`.
- Move `host` node in `mds_pim_print_core` configuration. 

## [1.1.0] - 2020-09-01
### Features
- Add UpdateElements command and selected elements filtering in `CommandQueue`.
- Add helper traits for concrete rendering services (E.g. `ElementCollectionRenderingTrait`).
- Add abstract template classes for standardized paper sizes.
- Add configurable form field to InDesign plugin.
- Set request and Pimcore locales to rendered language.
- Add extension possibilities to `AbstractPublicationTreeBuilder`.

### Breaking changes
- Change `GroupEnd` constructor signature.

### Fix
- Add configurable `LC_NUMERIC` locale setting to ensure InDesign compatible float to string conversion.

## [1.0.1] - 2020-08-06
### Changed
- Ensure float margin values in `PositionTrait`. 
- Update `mds-agenturgruppe/php-code-checker` to v1.0.1.

### Fixed
- BruteforceProtectionListener checks for InDesign requests.
 


src/DependencyInjection/Configuration.php
<?php
/**
 * mds PimPrint
 *
 * This source file is licensed under GNU General Public License version 3 (GPLv3).
 *
 * Full copyright and license information is available in
 * LICENSE.md which is distributed with this source code.
 *
 * @copyright  Copyright (c) mds. Agenturgruppe GmbH (https://www.mds.eu)
 * @license    https://pimprint.mds.eu/license GPLv3
 */

namespace Mds\PimPrint\CoreBundle\DependencyInjection;

use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;
use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

/**
 * Class Configuration
 *
 * @package Mds\PimPrint\CoreBundle\DependencyInjection
 */
class Configuration implements ConfigurationInterface
{
    /**
     * {@inheritDoc}
     *
     * @return TreeBuilder
     */
    public function getConfigTreeBuilder()
    {
        $treeBuilder = new TreeBuilder();
        $rootNode = $treeBuilder->root('mds_pim_print_core');
        $this->addProjectConfig($rootNode);

        return $treeBuilder;
    }

    /**
     * Add PimPrint projects configuration.
     *
     * @param ArrayNodeDefinition $rootNode
     */
    private function addProjectConfig(ArrayNodeDefinition $rootNode)
    {
        $rootNode->children()
                    ->arrayNode('lc_numeric')
                        ->info("Optional locales to set in render mode to have float values converted to '.-strings'")
                        ->example(['en_US.UTF-8', 'de_DE.UTF-8'])
                        ->prototype('scalar')->end()
                    ->end()
                    ->arrayNode('host')
                        ->info('Optional host settings for proxy or multi domain environments.')
                        ->children()
                            ->scalarNode('hostname')
                            ->info('Hostname')->end()
                            ->scalarNode('protocol')
                            ->info('Protocol')->end()
                            ->scalarNode('port')
                            ->info('Port')->end()
                        ->end()
                    ->end()
                    ->arrayNode('projects')
                    ->arrayPrototype()
                        ->children()
                            ->scalarNode('ident')
                                ->info('Internal project identifier.')->end()
                            ->scalarNode('name')->isRequired()
                                ->info('Name of project displayed in InDesign-Plugin.')->end()
                            ->scalarNode('service')->isRequired()
                                ->info('Service to use to render the project. Must inherit AbstractProject.')->end()
                            ->booleanNode('create_update_info')->defaultValue(true)
                                ->info('Toggles creation of update info layers.')->end()
                            ->arrayNode('template')->isRequired()
                                ->info('InDesign Template settings.')
                                ->children()
                                    ->scalarNode('default')
                                        ->info('Default InDesign template filename.')->end()
                                    ->scalarNode('relative_path')
                                        ->info('Optional relative path inside service bundle to the InDesign template.')
                                        ->defaultValue('/Resources/pimprint/')->end()
                                    ->booleanNode('download')->defaultValue(true)
                                        ->info('Download template')->end()
                                ->end()
                            ->end()
                            ->arrayNode('plugin_elements')->addDefaultsIfNotSet()
                                ->info('Available plugin elements.')
                                ->children()
                                    ->booleanNode('update_mode')->defaultValue(true)
                                        ->info('Field for update modes.')->end()
                                    ->arrayNode('update_modes')
                                        ->info('Available update modes for project.')
                                    ->defaultValue([501, 502, 512])
                                        ->prototype('scalar')->end()
                                    ->end()
                                    ->booleanNode('start_alignment')->defaultValue(false)
                                        ->info('Field for start left/right page.')->end()
                                    ->booleanNode('page_bounds')->defaultValue(false)
                                        ->info('Fields for page start/end.')->end()
                                ->end()
                            ->end()
                            ->arrayNode('assets')->addDefaultsIfNotSet()
                                ->info('Asset handling settings.')
                                ->children()
                                    ->booleanNode('download')->defaultValue(true)
                                        ->info('Toggles asset download.')->end()
                                    ->booleanNode('pre_download')->defaultValue(true)
                                        ->info('Toggles asset download before rendering start.')->end()
                                    ->booleanNode('warnings_on_page')->defaultValue(true)
                                        ->info('Toggles missing asset onPage warnings messages.')->end()
                                    ->scalarNode('thumbnail')
                                        ->info('Optional Pimcore thumbnail configuration for preview images.')->end()
                                ->end()
                            ->end()
                            ->scalarNode('php_time_limit')->defaultValue('0')
                                ->info('Optional PHP setting time_limit for project generation.')->end()
                            ->scalarNode('php_memory_limit')->defaultValue('2G')
                                    ->info('Optional PHP setting memory_limit for project generation.')->end()
                        ->end()
                    ->end()
                 ->end();
    }
}
